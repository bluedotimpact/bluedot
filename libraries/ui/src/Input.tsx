import clsx from 'clsx';
import { forwardRef } from 'react';

export type InputProps = {
  inputClassName?: string,
  label?: string,
  labelClassName?: string,
}
& React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>
& React.RefAttributes<HTMLInputElement>;

export const Input: React.ForwardRefExoticComponent<InputProps> = forwardRef(({
  className,
  inputClassName,
  labelClassName,
  ...props
}, ref) => {
  switch (props.type) {
    case 'checkbox':
      return (
        <label className={clsx('input flex items-center gap-2 cursor-pointer', labelClassName)}>
          <input
            {...props}
            ref={ref}
            className={clsx('input--checkbox size-6 accent-bluedot-normal rounded cursor-pointer', inputClassName)}
            type="checkbox"
          />
          <span className='input__label'>{props.label || props.value}</span>
        </label>
      );
    case 'radio':
      return (
        <label className={clsx('input flex items-center gap-2 cursor-pointer', labelClassName)}>
          <input
            {...props}
            ref={ref}
            className={clsx('input--radio size-6 accent-bluedot-normal rounded cursor-pointer', inputClassName)}
            type="radio"
          />
          <span className='input__label'>{props.label || props.value}</span>
        </label>
      );
    default:
      return (
        <label className={clsx('input flex flex-col gap-2', labelClassName)}>
          <span className='input__label'>{props.label}</span>
          <input
            {...props}
            ref={ref}
            className={clsx('input--text border border-color-divider rounded-lg p-4 focus-visible:outline focus-visible:outline-2 focus-visible:-outline-offset-2 focus-visible:outline-bluedot-normal', inputClassName)}
            type="text"
          />
        </label>
      );
  }
});
