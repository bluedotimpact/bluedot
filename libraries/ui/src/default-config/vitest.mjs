/* eslint-disable import/no-extraneous-dependencies */
import { defineConfig, defaultExclude } from 'vitest/config';
import react from '@vitejs/plugin-react';
import * as dotenv from 'dotenv';

/**
 * @param {import('vitest/config').UserConfig} [config] 
 * @returns {import('vitest/config').UserConfig}
 */
export const withDefaultBlueDotVitestConfig = async (config) => defineConfig({
  ...config,
  plugins: [react(), ...(config?.plugins ?? [])],
  test: {
    environment: 'happy-dom',
    env: dotenv.config({ path: '.env.test' }).parsed,

    ...config?.test,

    exclude: [...defaultExclude, '**/.{turbo,next}/**', ...(config?.test?.exclude ?? [])],
  },
});
