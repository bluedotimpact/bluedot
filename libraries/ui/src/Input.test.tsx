import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import { Input } from './Input';

describe('Input', () => {
  test('renders default as expected', () => {
    const { container } = render(
      <Input
        type="text"
        label="This is the label"
        placeholder="This is the placeholder"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('renders radio as expected', () => {
    const { container } = render(
      <Input
        type="radio"
        value="This is the value"
        name="radio-group"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('renders checkbox as expected', () => {
    const { container } = render(
      <Input
        type="checkbox"
        value="This is the value"
        name="checkbox-group"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
  