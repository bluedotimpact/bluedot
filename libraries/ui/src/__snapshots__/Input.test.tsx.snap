// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > renders checkbox as expected 1`] = `
<div>
  <label
    class="input flex gap-2 cursor-pointer has-[:disabled]:cursor-not-allowed items-center"
  >
    <input
      class="input--checkbox size-6 accent-bluedot-normal cursor-pointer disabled:cursor-not-allowed"
      name="checkbox-group"
      type="checkbox"
      value="This is the value"
    />
    <span
      class="input__label"
    >
      This is the value
    </span>
  </label>
</div>
`;

exports[`Input > renders default as expected 1`] = `
<div>
  <label
    class="input flex gap-2 cursor-pointer has-[:disabled]:cursor-not-allowed flex-col"
  >
    <span
      class="input__label"
    >
      This is the label
    </span>
    <input
      class="input--text border border-color-divider rounded-lg p-4 focus-visible:outline-2 focus-visible:-outline-offset-2 focus-visible:outline-bluedot-normal disabled:cursor-not-allowed disabled:bg-gray-50 disabled:text-gray-400"
      label="This is the label"
      placeholder="This is the placeholder"
      type="text"
    />
  </label>
</div>
`;

exports[`Input > renders radio as expected 1`] = `
<div>
  <label
    class="input flex gap-2 cursor-pointer has-[:disabled]:cursor-not-allowed items-center"
  >
    <input
      class="input--radio size-6 accent-bluedot-normal cursor-pointer disabled:cursor-not-allowed"
      name="radio-group"
      type="radio"
      value="This is the value"
    />
    <span
      class="input__label"
    >
      This is the value
    </span>
  </label>
</div>
`;
